Move ORA data download logic into ora2 repo.

dianakhuang opened the pull request:
Necessary for this ORA data download PR: https://github.com/edx/edx-platform/pull/11167

https://openedx.atlassian.net/browse/TNL-1402

dianakhuang commented:
@efischer19 and @robrap I believe this is ready for review.

efischer19 commented on openassessment/data.py:
```diff
@@ -333,4 +337,133 @@ def _use_read_replica(self, queryset):
+        if db.connection.vendor != 'mysql':
+            return '', ['']
+
+        raw_queries = cls.RAW_QUERY.split(';')
```
I think a comment here explaining why we have 2 different queries would be good

efischer19 commented on openassessment/data.py:
```diff
@@ -333,4 +337,133 @@ def _use_read_replica(self, queryset):
+        header = [item[0] for item in cursor.description]
+        return header, cursor.fetchall()
+
+    # Raw MySQL query that this CsvWriter
```
Incomplete comment?

efischer19 commented on openassessment/management/commands/collect_ora2_data.py:
```diff
@@ -0,0 +1,67 @@
+        if not args:
+            raise CommandError("Course ID must be specified to fetch data")
+
+        course_id = args[0]
```
Is there a way we can validate this near the start of the management command, and error out if it's not valid?

dianakhuang replied:
The problem is that ORA2 doesn't really have a concept of what a valid or invalid course id is. It's just a string in the code right now. I wasn't sure I wanted to add opaque-keys as a dependency just for this one minor use case.

efischer19 replied:
Ah, I see. Makes sense then.

efischer19 commented:
A few small issues, looks great overall. I definitely think we can get this into today's ORA release.

efischer19 commented:
:+1:

robrap commented on openassessment/data.py:
```diff
@@ -333,4 +337,135 @@ def _use_read_replica(self, queryset):
+    """
+
+    @classmethod
+    def _read_replica_cursor_if_available(cls):
```
Maybe call this `_read_replica_or_regular_cursor`?  The name as-is is slightly vague about what it returns in the case it is not available.

robrap commented on openassessment/data.py:
```diff
@@ -333,4 +337,135 @@ def _use_read_replica(self, queryset):
+
+        # Syntax unsupported by other vendors such as SQLite test db
+        if db.connection.vendor != 'mysql':
+            return '', ['']
```
Should this be an exception?

dianakhuang replied:
I'm not sure. That was the behavior in the original PR and I wasn't sure if there was anything that was relying on it.

robrap replied:
If we could determine and at least comment the mysql specific syntax, that might be helpful in the future.

robrap commented on openassessment/data.py:
```diff
@@ -333,4 +337,135 @@ def _use_read_replica(self, queryset):
+
+        # The raw query has two statements separated by a semi-colon.
+        # We would like to run them separately through the Django ORM.
+        raw_queries = cls.RAW_QUERY.split(';')
```
Any reason not to split RAW_QUERY into 2 variables with a reasonable name for each?

dianakhuang commented on openassessment/data.py:
```diff
@@ -333,4 +337,135 @@ def _use_read_replica(self, queryset):
+) AS `feedback`
+FROM `submissions_submission` AS `sub`
+JOIN `submissions_studentitem` AS `student` ON `sub`.`student_item_id`=`student`.`id`
+WHERE `student`.`item_type`="openassessment" AND `student`.`course_id`=%s
```
@ormsbee Here's the end of the giant query. Enjoy?

robrap replied:
Do we need to filter out data with status deleted from the submissions api, or should that already be properly handled by relying on valid ORA data?  You may want to test Clear Student State with the data download to verify.

efischer19 replied:
This depends on the use case - the whole point of doing soft-deletes was to keep the data available for later use. Using SQL directly will not filter out the deleted entries unless `submissions.status=D` is added to this query.

robrap commented on openassessment/data.py:
```diff
@@ -333,4 +337,135 @@ def _use_read_replica(self, queryset):
+    WHERE `feedback`.`submission_uuid`=`sub`.`uuid`
+    LIMIT 1
+) AS `feedback`
+FROM `submissions_submission` AS `sub`
```
We have the open question about whether or not we are ok with the fact that this query ignores the boundary between ORA and the Submissions API.  We discussed performance testing it as is to see if it is a non-issue (i.e. fails performance testing).

dianakhuang commented:
@robrap could you take a look at this and let me know what you think of the refactor work?

robrap commented on openassessment/data.py:
```diff
@@ -114,15 +115,15 @@ def write_to_csv(self, course_id):
                 AssessmentPart.objects.select_related('assessment', 'option', 'option__criterion')
-                    .filter(assessment__submission_uuid=submission_uuid)
-                    .order_by('assessment__pk')
+                .filter(assessment__submission_uuid=submission_uuid)
```
I personally prefer this as it was with the indent.  Do you mind showing me where we state to do otherwise, if we do?  Thanks.

dianakhuang replied:
This is a pep8 rule that we don't have exceptions for. I was just cleaning up the file on my way through.

robrap commented on openassessment/data.py:
```diff
@@ -333,4 +334,151 @@ def _use_read_replica(self, queryset):
+        )
+
+
+class AggregateOraData(object):
```
What do you think of ``OraAggregateData``?  The original tripped me up because it read like an "Or".  Also, it sounds a bit more like an action than a noun as it stands.

robrap commented on openassessment/data.py:
```diff
@@ -333,4 +334,151 @@ def _use_read_replica(self, queryset):
+    """
+    Aggregate all the ORA data into a single table.
+
+    Only works on MySQL databases.
```
MySQL only?  Should this comment be deleted?

dianakhuang commented on openassessment/data.py:
```diff
@@ -333,4 +334,149 @@ def _use_read_replica(self, queryset):
+
+class OraAggregateData(object):
+    """
+    Aggregate all the ORA data into a single table.
```
rename `table` to something clearer

dianakhuang commented on openassessment/data.py:
```diff
@@ -333,4 +334,149 @@ def _use_read_replica(self, queryset):
+        )
+
+    @classmethod
+    def _build_assessments_column(cls, assessments):
```
Rename column -> cell

dianakhuang commented:
@efischer19 This PR has been updated.

dianakhuang commented:
@robrap I updated those names.

efischer19 commented on openassessment/data.py:
```diff
@@ -333,4 +334,151 @@ def _use_read_replica(self, queryset):
+        Query MySQL database for aggregated ora2 response data.
+        """
+
+        all_submission_information = sub_api.get_all_course_submission_information(course_id, 'openassessment')
```
Note for myself so I can find the submissions integration point quickly when I need it again

efischer19 commented on openassessment/data.py:
```diff
@@ -333,4 +334,151 @@ def _use_read_replica(self, queryset):
+                )
+            )
+            assessments_cell = cls._build_assessments_cell(assessments)
+            assessments_parts_Cell = cls._build_assessments_parts_cell(assessments)
```
Inconsistent casing, can this be `assessments_parts_cell`?

efischer19 commented on openassessment/data.py:
```diff
@@ -333,4 +334,151 @@ def _use_read_replica(self, queryset):
+            returned_string += u"-- scorer_id: {}\n".format(assessment.scorer_id)
+            if assessment.feedback != u"":
+                returned_string += u"-- overall_feedback: {}\n".format(assessment.feedback)
+        return returned_string
```
:+1: I'm a fan of this system of including multiple assessments' data in a single cell in the resulting table.

efischer19 commented on openassessment/tests/test_data.py:
```diff
@@ -152,3 +225,108 @@ def _load_fixture(self, fixture_relpath):
+            'Feedback on Peer Assessments'
+        ])
+
+        self.assertEqual(data[1], [
```
Maybe I'm confused but why are we using `data[1]` here? Isn't `submission` created before `scorer_submission`? Or is the returned ordering not FIFO with respect to when the submissions were made?

efischer19 commented:
:+1: once we get the ordering issue sorted out, on to edx-submissions!

robrap commented on openassessment/data.py:
```diff
@@ -333,4 +334,152 @@ def _use_read_replica(self, queryset):
+    @classmethod
+    def collect_ora2_data(cls, course_id):
+        """
+        Query MySQL database for aggregated ora2 response data.
```
Anything MySQL specific left, or can this just say "Query database..." ?

robrap commented on openassessment/data.py:
```diff
@@ -333,4 +334,152 @@ def _use_read_replica(self, queryset):
+    def collect_ora2_data(cls, course_id):
+        """
+        Query MySQL database for aggregated ora2 response data.
+        """
```
Args and Returns?

robrap commented on openassessment/management/commands/collect_ora2_data.py:
```diff
@@ -0,0 +1,67 @@
+from optparse import make_option
+import os
+
+from openassessment.data import OraAggregateData
```
This import section (line) should be at the bottom.

robrap commented on openassessment/management/commands/collect_ora2_data.py:
```diff
@@ -0,0 +1,67 @@
+
+        course_id = args[0]
+
+        file_name = ("%s-ora2.csv" % course_id).replace("/", "-")
```
It feels like someone should be able to supply a file name to override this default.

robrap commented on openassessment/management/commands/collect_ora2_data.py:
```diff
@@ -0,0 +1,67 @@
+            writer.writerow(_preprocess(row))
+
+
+def _preprocess(data_list):
```
More specific:
```
def _encode_row(row):
```

robrap commented on openassessment/management/commands/collect_ora2_data.py:
```diff
@@ -0,0 +1,67 @@
+    """
+    Properly encode ora2 responses for transcription into a .csv
+    """
+    processed_data = []
```
processed_row (to go with above).

robrap commented on openassessment/management/tests/test_collect_ora2_data.py:
```diff
@@ -0,0 +1,71 @@
+            "feedback",
+        ]
+
+        self.test_row = [
```
Is this actually test_rows?  Also, I'd have 2 rows, otherwise same test would pass even if you didn't loop.

robrap commented on openassessment/management/tests/test_collect_ora2_data.py:
```diff
@@ -0,0 +1,71 @@
+            call_command('collect_ora2_data', self.COURSE_ID)
+
+            mock_writerow.assert_any_call(self.test_header)
+            mock_writerow.assert_called_with(self.test_row[0])
```
Seems like this is implicitly, but not explicitly testing the `unicode(item).encode('utf-8')`.  Should you add an assertion on the type of one of the items in the row before and after?  Does it require any special unicode characters in the test data as well?

robrap commented on openassessment/tests/test_data.py:
```diff
@@ -152,3 +225,124 @@ def _load_fixture(self, fixture_relpath):
+class TestOraAggregateData(TransactionCacheResetTest):
+    """
+    Test that the component parts of OraAggregateData behave as expected.
+    """
```
Maybe mention that this is an integration test, testing against the real submissions api.

robrap commented on openassessment/tests/test_data.py:
```diff
@@ -152,3 +225,124 @@ def _load_fixture(self, fixture_relpath):
+            u'',
+        ])
+
+        self.assertEqual(data[1], [
```
Seems like this doesn't give full code coverage.  Some examples:
- multiple assessments,
- without feedback
- without option label
There may be other cases.

robrap commented:
@dianakhuang Overall nice work.  Thank you.  It is clean, and above all, it separates the submissions api!  Just minor naming, commenting, testing feedback.

dianakhuang commented:
@robrap @efischer19 The last two commits should have the new tests and more addressing of code review feedback. When you get a chance, could you guys take another look?

robrap commented on openassessment/management/tests/test_collect_ora2_data.py:
```diff
@@ -0,0 +1,101 @@
+            mock_writerow.assert_any_call(self.test_rows[0])
+            mock_writerow.assert_any_call(self.unicode_encoded_row)
+
+            mock_writerow.mock_calls = []
```
1. I'm not following this.  It looks like the same test, with a different mock_writerow, and the same assertions for both tests.  
2. I was imagining a mock implementation of writerow which performs an assertion on the input to verify the list contains objects of type str.

robrap replied:
1. I'm not following this.  It looks like the same test, with a different mock_writerow, and the same assertions for both tests.  
2. I was imagining a mock implementation of writerow which performs an assertion on the input to verify the list contains objects of type str.

dianakhuang replied:
It is the same test, but now it's using two rows instead of one, and one of the rows contains unicode that we are checking is encoded properly. I don't know why the original test insisted on running the command twice, but I left it there in case there was a specific error condition they were checking for. I don't think it's there to check a different case.

robrap replied:
I forgot you didn't write the original test.  I'm almost thinking we should remove the second call_command if we don't see any good reason to have it, and no way to comment why it exists.

I saw the unicode encoded row.  I wasn't clear on how that would work with this assertion, but I think I see now.  Did you verify that if you stop the encoding inside encode_row() this test fails?

Thanks for explaining.

dianakhuang replied:
I just double-checked that pulling out `.encode('utf-8')` caused a failure in that line, and it does. I can remove the second `call_command` in the test because it really does seem redundant.

robrap commented:
Thanks for all the work @dianakhuang.  I didn't realize the mock code would be so extensive.  I appreciate the effort, and hopefully it could also get used for other tests in the future.

efischer19 commented:
Still :+1: from me

robrap commented:
:+1: Nice job.  Thank you.
